man/man3/folddata.3
.TH "folddata" 3 "28 Jul 2003" "VFML" \" -*- nroff -*-
.ad l
.nh
.SH NAME
folddata \- 
.SH "Detailed Description"
.PP 
Randomly splits a data set into a collection of train/test pairs. 

.SH "The folddata tool"
.PP
.PP
This tool is used by the \fBxvalidate\fP tool; you may find it useful.   For example, your algorithm might perform very poorly on a specific cross validation run, you could use this tool to reproduce the datasets and try to track down the problem.
.PP
Folddata splits a dataset into a number of testing and training sets as needed for doing cross-validation.  It takes each example in the original dataset and randomly assigns it to one of the 'folds' is it creating (note that this randomness means that the folds won't be exactly evenly sized).  Folddata then outputs one dataset for each fold with the examples from the fold as the test set and the examples in all the other folds as the training set.
.PP
Folddata works efficiently on large datasets, but will require enough disk space to hold 'folds' copies of the dataset.
.PP
Folddata takes input and does output in \fCc4.5 format\fP.   It expects to find the files \fC<stem>.names\fP and \fC<stem>.data\fP and outputs \fC<stem>[0 - n].[names, data, test]\fP.
.PP
.SS "Arguments"
.PP
.IP "\(bu" 2
-f <filestem>
.IP "  \(bu" 4
Set the stem name (default DF)
.PP

.IP "\(bu" 2
-target <dir>
.IP "  \(bu" 4
Set the output directory (default '.')
.PP

.IP "\(bu" 2
-source <dir>
.IP "  \(bu" 4
Set the directory that contains the dataset (default '.')
.PP

.IP "\(bu" 2
-folds <n>
.IP "  \(bu" 4
Sets the number of train/test sets to create (default 10)
.PP

.IP "\(bu" 2
-seed <n>
.IP "  \(bu" 4
Sets the random seed, multiple runs with the same seed will produce the same datasets (defaults to a random seed)
.PP

.IP "\(bu" 2
-h
.IP "  \(bu" 4
Display usage information and exit
.PP

.IP "\(bu" 2
-v
.IP "  \(bu" 4
Can be used multiple times to increase the debugging output
.PP

.PP
.PP
.SS "Example"
.PP
\fCfolddata -f test -target output -folds 15 -seed 10\fP
.PP
Will create 15 folds from test.names and test.data and put them in the directory named output as test[0-14].names, test[0-14].data, and test[0-14].test.  It will use a seeded random generator, so that the exact same dataset could be reproduced.
.PP
.SH SYNOPSIS
.br
.PP
.SH "Author"
.PP 
Generated automatically by Doxygen for VFML from the source code.


src/tools/folddata.c
#include "vfml.h"
#include <stdio.h>

//#include <string.h>

char  *gFileStem = "DF";
int   gFoldCount = 10;
int   gMessageLevel = 0;
char  *gTargetDirectory = ".";
char  *gSourceDirectory = ".";
long  gSeed = -1;
int   gError = 0;

static void _printUsage(char  *name) {
   printf("%s: splits a dataset into a collection of training and testing\n",
                                                                        name);
   printf("   sets for use with cross validation.  Writes the new datasets\n");
   printf("   as <stem>[0-n].<tail>\n");

   printf("-f <filestem>\tSet the name of the dataset (default DF)\n");
   printf("-target <dir>\tSet the output directory (default '.')\n");
   printf("-source <dir>\tSet the source data directory (default '.')\n");
   printf("-folds <n>\tSets the number of train/test sets to create (default 10)\n");
   printf("-seed <n>\tSets the random seed, multiple runs with the same seed will\n");
   printf("\t\t   produce the same datasets (defaults to a random seed)\n");

   printf("-v\t\tCan be used multiple times to increase the debugging output\n");
}

static void _processArgs(int argc, char *argv[]) {
   int i;

   /* HERE on the ones that use the next arg make sure it is there */
   for(i = 1 ; i < argc ; i++) {
      if(!strcmp(argv[i], "-f")) {
         gFileStem = argv[i+1];
         /* ignore the next argument */
         i++;
      } else if(!strcmp(argv[i], "-target")) {
         gTargetDirectory = argv[i+1];
         /* ignore the next argument */
         i++;
      } else if(!strcmp(argv[i], "-source")) {
         gSourceDirectory = argv[i+1];
         /* ignore the next argument */
         i++;
      } else if(!strcmp(argv[i], "-folds")) {
         sscanf(argv[i+1], "%d", &gFoldCount);
         /* ignore the next argument */
         i++;
      } else if(!strcmp(argv[i], "-v")) {
         gMessageLevel++;
      } else if(!strcmp(argv[i], "-seed")) {
         sscanf(argv[i+1], "%ld", &gSeed);
         /* ignore the next argument */
         i++;
      } else if(!strcmp(argv[i], "-h")) {
         _printUsage(argv[0]);
         exit(0);
      } else {
         printf("Unknown argument: %s.  use -h for help\n", argv[i]);
         exit(0);
      }
   }

   if(gMessageLevel >= 1) {
      printf("File stem: %s\n", gFileStem);
      printf("Target: %s\n", gTargetDirectory);
      printf("Source: %s\n", gSourceDirectory);
      printf("Folds: %d\n", gFoldCount);
      printf("Seed: %ld\n", gSeed);
   }
}

int main(int argc, char *argv[]) {
   char fileName[255];
   int i, targetFold;
   FILE *exampleIn, *names;
   ExampleSpecPtr es;
   ExamplePtr e;
   FILE **outputData;
   FILE **outputTest;
   long *outputCounts;

   _processArgs(argc, argv);

   /* set up the random seed */
   RandomInit();
   if(gSeed != -1) {
      RandomSeed(gSeed);
   }

   /* read the spec file */
   sprintf(fileName, "%s/%s.names", gSourceDirectory, gFileStem);
   /* TODO check that the file exists */
   es = ExampleSpecRead(fileName);
   DebugError(es == 0, "Unable to open the .names file");


   /* set up the output files */
   outputData = MNewPtr(sizeof(FILE *) * gFoldCount);
   outputTest = MNewPtr(sizeof(FILE *) * gFoldCount);
   outputCounts = MNewPtr(sizeof(long) * gFoldCount);

   for(i = 0 ; i < gFoldCount ; i++) {
      /* TODO check that the target directory exists */
      sprintf(fileName, "%s/%s%d.data", gTargetDirectory, gFileStem, i);
      outputData[i] = fopen(fileName, "w");
      sprintf(fileName, "%s/%s%d.test", gTargetDirectory, gFileStem, i);
      outputTest[i] = fopen(fileName, "w");
      outputCounts[i] = 0;

      DebugError(outputData[i] == 0 || outputTest[i] == 0, "Unable to open output files.");
   }

   sprintf(fileName, "%s/%s.data", gSourceDirectory, gFileStem);
   exampleIn = fopen(fileName, "r");
   DebugError(exampleIn == 0, "Unable to open .data file");

   e = ExampleRead(exampleIn, es);
   while(e != 0) {
      /* DO THE FOLDING */
      targetFold = RandomRange(0, gFoldCount - 1);
      for(i = 0 ; i < gFoldCount ; i++) {
         if(i == targetFold) {
            ExampleWrite(e, outputTest[i]);
            outputCounts[i]++;
         } else {
            ExampleWrite(e, outputData[i]);
	 }
      }

      ExampleFree(e);
      e = ExampleRead(exampleIn, es);
   }
   fclose(exampleIn);

   /* close the output files */
   for(i = 0 ; i < gFoldCount ; i++) {
      /* TODO check that the target directory exists */
      fclose(outputData[i]);
      fclose(outputTest[i]);
      if(gMessageLevel >= 1) {
         printf("fold %d: %ld\n", i, outputCounts[i]);
      }
   }

   /* output the names files */
   for(i = 0 ; i < gFoldCount ; i++) {
      /* TODO check that the target directory exists */
      sprintf(fileName, "%s/%s%d.names", gTargetDirectory, gFileStem, i);
      names = fopen(fileName, "w");
      DebugError(names == 0, "Unable to open target names file for output.");

      ExampleSpecWrite(es, names);
      fclose(names);
   }

   return 0;
}

